File: Animations/Sunrays/C/effect.c
#include "def.h"

// --------------------------------------------------------------------------------------------------------------------

void calculateColors(EffectManager *effect)
{
    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        Color color = ColorFromHSV((float)effect->particles[i].x + effect->particles[i].y, 1.0, 1.0);
        effect->colors[i] = color;
    }
}

// --------------------------------------------------------------------------------------------------------------------

void createParticles(EffectManager *effect, Config *config)
{
    Particle *particleArray = (Particle *)malloc(sizeof(Particle) * effect->totalParticles + 1);
    effect->particles = particleArray;

    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        Particle newParticle = {
            radius : GetRandomValue(5, 15),
            friction : 0.95,
            pushX : 0,
            pushY : 0,
        };
        effect->particles[i] = newParticle;
    }

    setPosition(effect, config);
}

// --------------------------------------------------------------------------------------------------------------------

void draw(EffectManager *effect, Mouse *mouse)
{
    Color color = {255, 255, 255, 255 * 0.4};
    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        if (i % effect->rayDivisor == 0)
        {
            Vector2 start = {effect->particles[i].x, effect->particles[i].y};
            DrawLineEx(start, mouse->mouseVector2, 1.0, color);
        }
    }

    for (int32_t i = 0; i < effect->lines.total; i++)
    {
        DrawLineEx(effect->lines.start[i], effect->lines.end[i], effect->lineThickness, effect->lines.color[i]);
    }

    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        DrawCircle(effect->particles[i].x, effect->particles[i].y, (float)effect->particles[i].radius, effect->colors[i]);
    }
}

// --------------------------------------------------------------------------------------------------------------------

void findConnections(EffectManager *effect)
{
    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        for (int32_t j = i; j < effect->totalParticles; j++)
        {
            if (i == j)
            {
                continue;
            }

            int32_t dx = effect->particles[i].x - effect->particles[j].x;
            int32_t dy = effect->particles[i].y - effect->particles[j].y;
            double distance = hypot(dx, dy);

            if (distance < effect->connectionDistance)
            {
                Vector2 start = {effect->particles[i].x,
                                 effect->particles[i].y};
                Vector2 end = {effect->particles[j].x,
                               effect->particles[j].y};

                double opacity = 1 - (distance / effect->connectionDistance);
                Color color = {
                    255,
                    255,
                    255,
                    255 * opacity,
                };

                effect->lines.start[effect->lines.total] = start;
                effect->lines.end[effect->lines.total] = end;
                effect->lines.color[effect->lines.total] = color;
                effect->lines.total++;

                checkAllocationLines(effect);
            }
        }
    }
}

// --------------------------------------------------------------------------------------------------------------------

void setPosition(EffectManager *effect, Config *config)
{
    for (int i = 0; i < effect->totalParticles; i++)
    {
        effect->particles[i].x = GetRandomValue(effect->particles[i].radius, config->width - effect->particles[i].radius);
        effect->particles[i].y = GetRandomValue(effect->particles[i].radius, config->height - effect->particles[i].radius);
        effect->particles[i].movementX = GetRandomValue(-2, 2);
        effect->particles[i].movementY = GetRandomValue(-2, 2);
    }
}

// --------------------------------------------------------------------------------------------------------------------

void updateEffect(EffectManager *effect, Mouse *mouse, Config *config)
{

    for (int32_t i = 0; i < effect->totalParticles; i++)
    {
        if (IsMouseButtonDown(MOUSE_BUTTON_LEFT))
        {
            updateMouse(mouse);
            pushParticle(&effect->particles[i], mouse);
        }

        moveParticle(&effect->particles[i]);
        collision(&effect->particles[i], config);
    }

    findConnections(effect);
    calculateColors(effect);
}

// --------------------------------------------------------------------------------------------------------------------

File: Animations/Sunrays/Python/effect.py
import math
from mouse import Mouse
from particle import Particle
import pyray as rl


class Effect:
    def __init__(
        self,
        window_width: int,
        window_height: int,
        total_particles: int,
        link_distance: int,
        line_thickness: float,
        mouse_radius: int,
    ) -> None:
        self.window_width = window_width
        self.window_height = window_height
        self.total_particles = total_particles
        self.particles: list[Particle] = []
        self.create_particles()

        self.link_distance = link_distance
        self.line_thickness = line_thickness
        self.line_start: list[rl.Vector2] = []
        self.line_end: list[rl.Vector2] = []
        self.line_color: list[rl.Color] = []

        self.mouse = Mouse(mouse_radius)

    # ------------------------------------------------------------------------------------------------------------------

    def calculate_connections(self) -> None:
        for i, _ in enumerate(self.particles):
            for j in range(i, len(self.particles)):
                if i == j:
                    continue
                dx = self.particles[i].x - self.particles[j].x
                dy = self.particles[i].y - self.particles[j].y
                distance = math.hypot(dx, dy)

                if distance < self.link_distance:
                    self.line_start.append(
                        rl.Vector2(self.particles[i].x, self.particles[i].y)
                    )
                    self.line_end.append(
                        rl.Vector2(self.particles[j].x, self.particles[j].y)
                    )
                    opacity = 1 - (distance / self.link_distance)
                    self.line_color.append(rl.Color(255, 255, 255, int(255 * opacity)))

    # ------------------------------------------------------------------------------------------------------------------

    def create_particles(self) -> None:
        for _ in range(self.total_particles):
            self.particles.append(Particle(self.window_width, self.window_height))

    # ------------------------------------------------------------------------------------------------------------------

    def draw_particles(self) -> None:
        for i, line in enumerate(self.line_start):
            rl.draw_line_ex(
                self.line_start[i],
                self.line_end[i],
                self.line_thickness,
                self.line_color[i],
            )

        for particle in self.particles:
            particle.draw()

    # ------------------------------------------------------------------------------------------------------------------

    def resize_screen(self, new_width: int, new_height: int) -> None:
        self.window_width = new_width
        self.window_height = new_height

        for particle in self.particles:
            particle.window_width = self.window_width
            particle.window_height = self.window_height
            particle.set_initial_values()

    # ------------------------------------------------------------------------------------------------------------------

    def update_particles(self) -> None:
        self.line_start = []
        self.line_end = []
        self.line_color = []

        self.calculate_connections()

        self.mouse.update_pressed()
        if self.mouse.pressed:
            self.mouse.update()

        for particle in self.particles:
            particle.update(
                self.mouse.x, self.mouse.y, self.mouse.radius, self.mouse.pressed
            )

    # ------------------------------------------------------------------------------------------------------------------
